### PR TITLE
automatically switch to vertical mode when needed

### LINKED ISSUE #292: Default orientation based on window dimensions
Add an "Auto-switch to vertical mode when appropriate" option

### PR DESCRIPTION
**Pull request by KentoNishi**
Added an `Override layout settings when entering vertical mode` setting to handle orientation changes automatically. Most useful on Android, but could also apply to tablets. Resolves https://github.com/LiveTL/LiveTL/issues/292

https://user-images.githubusercontent.com/38841491/128476788-16689350-c2f0-4b07-a764-22a9e4411448.mp4

### CODE REVIEW

File: src/components/Watch.svelte [hunk 1]
```diff
@@ -86,6 +92,14 @@
     });
   };
   $: setTimeout(() => changeSide($videoSide, $chatSplit), 0);
+
+  function orientationChangeCallback() {
+    $videoSide = ($autoVertical && window.innerHeight > window.innerWidth) ? VideoSide.TOP : $videoSideSetting;
+  }
+  window.addEventListener('resize', orientationChangeCallback);
+  videoSideSetting.subscribe(orientationChangeCallback);
+  autoVertical.subscribe(orientationChangeCallback);
```
Comment (ChrRubin) [inline src/components/Watch.svelte]: Should unsubscribe at onDestroy.

Or even better, replace it with `$: $videoSideSetting, $autoVertical, orientationChangeCallback();`, which would call the function whenever any of the auto-subscribe stores updates.

**Review by ChrRubin — CHANGES_REQUESTED** [review]

File: src/components/Watch.svelte [hunk 2]
```diff
@@ -86,6 +92,14 @@
     });
   };
   $: setTimeout(() => changeSide($videoSide, $chatSplit), 0);
+
+  function orientationChangeCallback() {
+    $videoSide = ($autoVertical && window.innerHeight > window.innerWidth) ? VideoSide.TOP : $videoSideSetting;
+  }
+  window.addEventListener('resize', orientationChangeCallback);
```
Comment (r2dev2) [inline src/components/Watch.svelte]: Use ``<svelte:window on:resize />`` to make it auto-clean-up the ``orientationChange`` callback

File: src/components/Watch.svelte (hunk 1 above)
Comment (r2dev2) [reply src/components/Watch.svelte]: You could also make a ``changeOrientation`` function and do
```js
$: changeOrientation($videoSideSetting, $autoVertical)
```

Then you can call this function in the orientationChangeCallback

File: src/components/Watch.svelte [hunk 3]
```diff
@@ -25,6 +27,10 @@
   import ChatEmbed from './ChatEmbed.svelte';
   import Popout from './Popout.svelte';
   import Captions from './Captions.svelte';
+  import {
+    onMount,
+  } from 'svelte';
+
```
Comment (r2dev2) [inline src/components/Watch.svelte]: Why this need to be multiple line import lmao

File: src/components/settings/UISettings.svelte [hunk 4]
```diff
@@ -22,6 +24,8 @@
   import FontDemo from '../FontDemo.svelte';
   import ImportExport from '../ImportExport.svelte';
   export let isStandalone = false;
+
+  videoSideSetting.subscribe(value => $videoSide = value);
```
Comment (r2dev2) [inline src/components/settings/UISettings.svelte]: Move this to ``js/store.js`` since it's not guaranteed that this code will run.

File: src/components/Watch.svelte [hunk 5]
```diff
@@ -86,6 +92,14 @@
     });
   };
   $: setTimeout(() => changeSide($videoSide, $chatSplit), 0);
+
+  function orientationChangeCallback() {
+    $videoSide = ($autoVertical && window.innerHeight > window.innerWidth) ? VideoSide.TOP : $videoSideSetting;
+  }
+  window.addEventListener('resize', orientationChangeCallback);
+  videoSideSetting.subscribe(orientationChangeCallback);
```
Comment (r2dev2) [inline src/components/Watch.svelte]: ``videoSideSetting`` is also subscribed in ``components/UISettings.svelte`` to set the value of ``videoSide`` whenever it changes. I'm confused.

File: src/components/Watch.svelte [hunk 6]
```diff
@@ -86,6 +92,14 @@
     });
   };
   $: setTimeout(() => changeSide($videoSide, $chatSplit), 0);
+
+  function orientationChangeCallback() {
+    $videoSide = ($autoVertical && window.innerHeight > window.innerWidth) ? VideoSide.TOP : $videoSideSetting;
```
Comment (r2dev2) [inline src/components/Watch.svelte]: This logic to change ``videoSide`` which is distributed across multiple files is getting really confusing. You should refactor the logic of ``videoSide`` and turn it into a [readable store](https://svelte.dev/tutorial/readable-stores) so all the mutations are managed at one place and we know what's happening.

**Review by r2dev2 — CHANGES_REQUESTED** [review]

**Comment by r2dev2** [comment]
TLDR: use readable store for ``videoSideSetting`` to make the logic more understandable.

**Comment by r2dev2** [comment]
> TLDR: use readable store for `videoSideSetting` to make the logic more understandable.

Edit: You may find more success with this by wrapping ``window.addEventListener('resize')`` with a ``readable`` store and then making ``videoSideSetting`` a ``derived`` store.

**Comment by KentoNishi** [comment]
now that i think about it, readable store is indeed better, refactor 2 incoming

File: src/components/Watch.svelte [hunk 7]
```diff
@@ -25,6 +25,8 @@
   import ChatEmbed from './ChatEmbed.svelte';
   import Popout from './Popout.svelte';
   import Captions from './Captions.svelte';
+  import { onMount } from 'svelte';
+
```
Comment (ChrRubin) [inline src/components/Watch.svelte]: Unused import now that you're using readable.

File: src/components/settings/UISettings.svelte [hunk 8]
```diff
@@ -40,8 +41,14 @@
...
       options={Object.keys(VideoSide)}
-      store={videoSide}
+      store={videoSideSetting}
     />
+    <div style="margin-bottom: 10px;">
+      <CheckOption
+        name="Override layout settings when entering vertical mode"
+        store={autoVertical}
+      />
+    </div>
```
Comment (ChrRubin) [inline src/components/settings/UISettings.svelte]: I'm wondering if there's a better way to word these, cause it might be confusing for the video side setting to show left/right even when it's being overwritten during vertical mode. 

I guess for starters "Override layout settings" can be changed to "Override video side" since the former suggests multiple settings will be overridden.

File: src/js/store.js [hunk 9]
```diff
@@ -72,11 +72,29 @@ export const
...
+    unsubAutoVerticalListener();
+    unsubVideoSideSettingListener();
+  };
+});
 export const updatePopupActive = writable(false);
 export const videoTitle = writable('LiveTL');
 export const timestamp = writable(0);
 export const faviconURL = writable('/48x48.png');
-export const availableMchadUsers = writable([]);
+export const availableMchadUsers = writable([]);
```
Comment (ChrRubin) [inline src/js/store.js]: Slight pet peeve of mine not having newline at the end of file lmao

**Review by ChrRubin — CHANGES_REQUESTED** [review]

File: src/components/settings/UISettings.svelte (hunk 8 above)
Comment (r2dev2) [reply src/components/settings/UISettings.svelte]: Maybe it could be named `Enter vertical mode when window is thin`?

**Review by r2dev2 — COMMENTED** [review]
Looks good except for the naming of the setting.

**Comment by KentoNishi** [comment]
i'll merge after u give it one final look

**Review by r2dev2 — APPROVED** [review]
